src/libutil/url.cc

```diff
@@ -308,10 +336,43 @@ std::string encodeQuery(const StringMap & ss)
     return res;
 }
 
+CanonPath urlPathToCanonPath(const std::vector<std::string> & urlPath)
+{
+    for (const auto & comp : urlPath) {
+        if (comp.find('/') != comp.npos) {
+            throw Error("URL path component contains '/', which is not allowed in file names");
+        }
+    }
+
+    return CanonPath(urlPath);
+}
+
+// base case
+static inline auto sum(auto && first)
+{
+    return first;
+}
+
+// multiple parameters case
+static inline auto sum(auto && first, auto &&... rest)
+{
+    return first + sum(rest...);
+}
```

Comment:
Can you instead do something like:

```suggestion
static inline auto sum(auto &&... rest)
{
    std::string res;
    ((res += rest), ...);
    return res;
}
```

Reply:
oh, very nice :)